### ISSUE TITLE
WebTransport.ready

### ISSUE THREAD

**Issue by yutakahirano** [w3c/webtransport#92]
Like #81, I think it's good to have `.ready` property and remove `.onstatechange` and `.state` from https://wicg.github.io/web-transport/#web-transport.

**Comment by lgrahl** [comment]
I don't think this is a good idea. First of all, states can change. It makes sense to expose them via an event. Secondly, you can directly wait for whatever you want to wait for by using the promises of the read/write streams.

**Comment by ricea** [comment]
There are only a small number of state transitions that are possible:
```
connecting -> connected
connecting -> failed
connected -> closed
connected -> failed
```
All of these can be reliably detected using a combination of the `.ready` and `.closed` promises.

The trouble with leaving `.onstatechange` there "just in case" is that people will use it because it's there. This will lead to buggy XHR-style code that doesn't handle all the state transitions correctly.

Similarly `.state` implies a synchronous model which doesn't really reflect the way that modern browsers work.

I am trying to make the first version of the API as minimal as possible, because it is far easier to add things afterwards than it is to take them away.

**Comment by lgrahl** [comment]
> I am trying to make the first version of the API as minimal as possible, because it is far easier to add things afterwards than it is to take them away.

What if we come to the conclusion that we need more states? We've been there before: https://github.com/w3c/webrtc-pc/issues/1827
Do we add more promises in that case? That would make it quite convoluted. I already find two promises (`.ready` and `.closed`) quite clunky and error-prone (due to potential races). IMO a single source for state events is a good thing.

Also, again, what is the problem with awaiting the underlying read/write streams?

**Comment by yutakahirano** [comment]
>> I am trying to make the first version of the API as minimal as possible, because it is far easier to add things afterwards than it is to take them away.

> What if we come to the conclusion that we need more states? We've been there before: w3c/webrtc-pc#1827
> Do we add more promises in that case? That would make it quite convoluted.

That depends of the change, but we may need to. I'm not sure if it would be quite convoluted.

This is WebTransportState, so we don't want to have transport-specific states here. If we need to do that, more drastic change would be needed so I don't think we'll have many states with maintaining the current mechanism.

Synchronous event handling is difficult to use / difficult to implement (e.g, reentrancy) so I would like to avoid that if possible.

> Also, again, what is the problem with awaiting the underlying read/write streams?

I'm not sure I understand this question. This is connection's state and streams states are different from connection's state.

**Comment by lgrahl** [comment]
> Synchronous event handling is difficult to use / difficult to implement (e.g, reentrancy) so I would like to avoid that if possible.

The dispatching from the native transport doesn't have to be synchronous and probably should not be synchronous; as long as there is a clear order on how the state is bubbled up and to where because it eventually has to land in the in/out streams as well (when it's being closed). I'd suggest clearly defining that order.

Whether handling of the state event within JS is synchronous is up to the application and that would be the same with promises. Or is there a difference I'm not seeing?

> I'm not sure I understand this question. This is connection's state and streams states are different from connection's state.

Yes, but what I'd assume 99% of use cases care about is whether or not they can send/receive data. In any case, I'd suggest either removing the `closed` promise or removing the state events. Having both seems like a footgun.

---

Another way of getting rid of these state events would be to return the transport within a `Promise<WebTransport>` since there's nothing to do with the transport as long as it's not connected. In that case, the `closed` promise would suffice for both normal closure and failure since it could reject in case of going into the `failure` state.

Clarifying:

```webidl
interface QuicTransport {
  static Promise<QuicTransport> connect(DOMString host, unsigned short port);
};

QuicTransport includes UnidirectionalStreamsTransport;
QuicTransport includes BidirectionalStreamsTransport;
QuicTransport includes DatagramTransport;
QuicTransport includes WebTransport;

interface mixin WebTransport {
  readonly attribute Promise<WebTransportCloseInfo> closed;
  void close(optional WebTransportCloseInfo closeInfo);
);
```

Would :heart: that.

**Comment by ricea** [comment]
> Another way of getting rid of these state events would be to return the transport within a Promise<WebTransport> since there's nothing to do with the transport as long as it's not connected.

While I like the idea of only exposing the methods once connection is done, it would increase the number of non-constructible classes in the specification. We've already got too many violations of the ["Classes should have constructors when possible"](https://w3ctag.github.io/design-principles/#constructors) design principle. We should be careful about adding more.

**Comment by lgrahl** [comment]
Well, as you've said, this API already heavily uses mixins. Furthermore, the TAG group document states that

> Sometimes this isn’t the right answer

And what can one do with an unconnected transport instance that doesn't expose anything useful until connected? It would be awkward to stop now where it really brings an advantage.

**Comment by murillo128** [comment]
I think we should not remove the events (or make it observable at least) and I think we should add a promise based method for awaiting on it before doing anything on transport. 

Also note that the `.state` and `.onstatechange` are members of `WebTransport` and not from `QuicTransport`. Nothing prevents other transports to have a more complex state machine. Where do you propose to add the `.ready()` method?

**Comment by yutakahirano** [comment]
> Where do you propose to add the .ready() method?

On WebTransport.

> Also note that the .state and .onstatechange are members of WebTransport and not from QuicTransport. Nothing prevents other transports to have a more complex state machine.

If we keep `.state` and `.onstatechange` that would prevent individual protocols from introducing them.

**Comment by jan-ivar** [comment]
WebTransport TPAC meeting notes
* Consensus to add .ready promise
* Open new issue on deprecating state and onstatechange
   * Today's states cannot change more than once.
   * zero-rtt? a state?
   * Ask guidance from the TAG
